Added MQTTv5 UTransport methods

matthewd0123 opened the pull request:
* Added UTransport (send, registerlistener, unregisterlistener) methods
* Added pub/sub examples for testing.
* Updated README with information on installation and use.
* Added necessary legal/contributing documents
* Added publish workflow

matthewd0123 committed:
Added MQTTv5 UTransport methods

* Added UTransport (send, registerlistener, unregisterlistener) methods
* Added pub/sub examples for testing.
* Updated README with information on installation and use.
* Added necessary legal/contributing documents
* Added publish workflow

matthewd0123 commented:
#2

devkelley commented on README.adoc:
```diff
@@ -0,0 +1,94 @@
+---
+
+=== Using UTransport APIs
+
```
In this scenario you have this listed as `mqtt5_subscriber`. Would an `mqtt5_subscriber` be initialized the same way as it is done in `Creating a Instance of MQTT5UTransport`?

devkelley commented on up_client_mqtt5_python/mqtt5_utransport.py:
```diff
@@ -0,0 +1,303 @@
+        #     # 7: UPAYLOAD_FORMAT_TEXT
+        #     payload = base64.b64encode(payload_proto.data)
+
+        payload: bytes = payload_proto.SerializeToString()
```
The spec still needs to determine how we are serializing the protobuf payload to bytes, but I think this is fine for now

matthewd0123 committed:
Add Handlers for all UAttributes

All Properties in UAttributes will be mapped to MQTT Publish Properties. As Such, I have added this into the UTransport.

nb: I am using the in-built Base64ProtobufSerializer() methods to serialize and deserialize the UUIDs (id and reqid). This implementation may change.

matthewd0123 committed:
Add Test Cases, Support for UserProperties

* Testing Send and RegisterListener using paho.mqtt FakeBroker and Pytest
* Support all userproperties mapping to UAttributes

matthewd0123 committed:
Few changes related to formatting

matthewd0123 committed:
Modify licenses to match requirements

devkelley commented on up_client_mqtt5_python/mqtt5_utransport.py:
```diff
@@ -0,0 +1,396 @@
+
+        return UStatus(code=UCode.OK, message="OK")
+
+    def register_listener(self, topic: UUri, listener: UListener) -> UStatus:
```
I believe that there is a PR currently out for 1.5.8 that updates the register_listener, unregister_listener and receive functions to use a source_filter and a sink_filter. @sophokles73 and @stevenhartley, is this correct that the uTransport spec change is expected to get in for 1.5.8?

sophokles73 replied:
yes, the corresponding PR is https://github.com/eclipse-uprotocol/up-spec/pull/134 which will go into 1.5.8

devkelley commented on tests/testsupport/mqtt5_props.py:
```diff
@@ -0,0 +1,100 @@
+# Copyright (C) 2021
```
This file and some other files in test have this license. Is this the correct license?

sophokles73 replied:
No, it refers to the EPL but uProtocol uses the Apache Software License 2.0. It should be
```
# SPDX-FileCopyrightText: 2021 Contributors to the Eclipse Foundation
#
# See the NOTICE file(s) distributed with this work for additional
# information regarding copyright ownership.
#
# This program and the accompanying materials are made available under the
# terms of the Apache License Version 2.0 which is available at
# https://www.apache.org/licenses/LICENSE-2.0
#
# SPDX-License-Identifier: Apache-2.0
```

matthewd0123 replied:
@devkelley @sophokles73 for some of the files in tests folder, I have brought in test code from the paho.mqtt.python project, which is an eclipse project (https://github.com/eclipse/paho.mqtt.python/tree/master/tests). I have used the licensing that they used in some of their files. Can one of you let me know if there is a different way to reference that I'm using their code, or if it's even allowed to take code from other eclipse projects?

devkelley replied:
Hmm, that is a good question. I am sure it depends partly on the EPL-2.0 license and what that stipulates. This FAQ seems to imply that it is okay to include as long as you include where you got the code from and the license (https://www.eclipse.org/legal/epl-2.0/faq.php). That being said it would probably be good to verify with someone from eclipse foundation

matthewd0123 committed:
Update to 1.5.8, Formatting

* Update MQTT Client and examples to 1.5.8
* Format to align with ruff standards
* Add linter to workflow

matthewd0123 committed:
Update examples to work with Java MQTT

matthewd0123 committed:
Update up-python version to 0.2.0

matthewd0123 committed:
Update Message Handler to support wildcards

sophokles73 commented on .github/ruff.toml:
```diff
@@ -0,0 +1,37 @@
+quote-style = "preserve"
+indent-style = "space"
+docstring-code-format = true
+docstring-code-line-length = 100
```
missing EOL

matthewd0123 replied:
resolved all missing EOLs you pointed out, thanks!

sophokles73 commented on .github/workflows/lint.yml:
```diff
@@ -0,0 +1,43 @@
+
+      - name: Check Python Code Formatting with Ruff
+        run: |
+          ruff format --check --config .github/ruff.toml
```
missing EOL

sophokles73 commented on .github/workflows/python-publish.yml:
```diff
@@ -0,0 +1,49 @@
+      uses: pypa/gh-action-pypi-publish@27b31702a0e7fc50959f5ad993c78deac1bdfc29
+      with:
+        user: __token__
+        password: ${{ secrets.PYPI_TOKEN }}
```
missing EOL

sophokles73 commented on .gitignore:
```diff
@@ -0,0 +1,3 @@
+.vscode/**
+**/.coverage
+**/__pycache__
```
missing EOL

sophokles73 commented on examples/docker_mosquitto/docker-compose.yml:
```diff
@@ -0,0 +1,43 @@
+        target: /mosquitto/data
+      - type: bind
+        source: ./mosquitto/log
+        target: /mosquitto/log
```
missing EOL

sophokles73 commented on examples/docker_mosquitto/mosquitto/config/mosquitto.conf:
```diff
@@ -0,0 +1,36 @@
+log_timestamp true
+log_timestamp_format %Y-%m-%dT%H:%M:%S
+
+max_queued_messages 0
```
missing EOL

matthewd0123 committed:
Address Comments

* Add EOL to files in repository that were missing them.
